concept/type_/constraint.rs

```diff
@@ -4,139 +4,544 @@
+        );
+        debug_assert!(
+            &constraint.source() == &source,
+            "Unexpected different source of an abstract constraint for type"
+        );
+        Some(constraint.clone())
+    } else {
+        None
+    }
+}
+
+pub(crate) fn get_unique_constraints<C: Constraint<T>, T: Hash + Eq>(
+    constraints: impl IntoIterator<Item = C>,
+) -> HashSet<C> {
+    filter_by_constraint_category!(constraints.into_iter(), Unique).collect()
+}
+
+pub(crate) fn get_unique_constraint<'a, C: Constraint<T> + 'a, T: Hash + Eq>(
+    constraints: impl IntoIterator<Item = &'a C>,
+) -> Option<C> {
+    let mut uniques = filter_by_constraint_category!(constraints.into_iter(), Unique).into_iter();
+    if let Some(constraint) = uniques.next() {
+        debug_assert!(uniques.next().is_none(), "Expected to inherit only one unique constraint from its root source");
+        Some(constraint.clone())
+    } else {
+        None
+    }
+}
+
+pub(crate) fn get_distinct_constraints<C: Constraint<T>, T: Hash + Eq>(
+    constraints: impl IntoIterator<Item = C>,
+) -> HashSet<C> {
+    filter_by_constraint_category!(constraints.into_iter(), Distinct).collect()
+}
+
+pub(crate) fn get_independent_constraints<C: Constraint<T>, T: Hash + Eq>(
+    constraints: impl IntoIterator<Item = C>,
+) -> HashSet<C> {
+    filter_by_constraint_category!(constraints.into_iter(), Independent).collect()
+}
+
+pub(crate) fn get_regex_constraints<C: Constraint<T>, T: Hash + Eq>(
+    constraints: impl IntoIterator<Item = C>,
+) -> HashSet<C> {
+    filter_by_constraint_category!(constraints.into_iter(), Regex).collect()
+}
+
+pub(crate) fn get_range_constraints<C: Constraint<T>, T: Hash + Eq>(
+    constraints: impl IntoIterator<Item = C>,
+) -> HashSet<C> {
+    filter_by_constraint_category!(constraints.into_iter(), Range).collect()
+}
+
+pub(crate) fn get_values_constraints<C: Constraint<T>, T: Hash + Eq>(
+    constraints: impl IntoIterator<Item = C>,
+) -> HashSet<C> {
+    filter_by_constraint_category!(constraints.into_iter(), Values).collect()
+}
+
+pub(crate) fn get_checked_constraints<C: Constraint<T>, T: Hash + Eq>(
+    constraints: impl IntoIterator<Item = C>,
+) -> HashSet<C> {
+    filter_out_unchecked_constraints!(constraints.into_iter()).collect()
+}
+
+pub(crate) fn get_owns_default_constraints<CAP: Capability<'static>>(
+    source: CAP,
+    ordering: Ordering,
+) -> HashSet<CapabilityConstraint<CAP>> {
+    let mut constraints = HashSet::from([CapabilityConstraint::new(
+        ConstraintDescription::Cardinality(Owns::get_default_cardinality(ordering)),
+        source.clone(),
+    )]);
+
+    if let Some(default_distinct) = Owns::get_default_distinct(ordering) {
+        constraints.insert(CapabilityConstraint::new(ConstraintDescription::Distinct(default_distinct), source));
+    }
+
+    constraints
+}
+
+pub(crate) fn get_plays_default_constraints<CAP: Capability<'static>>(
```

Comment:
For simplicity, unordered `owns` and `relates` have a `distinct` constraint (so we could check "if distinct then..." without worrying about ordering). I wasn't sure what to do with plays:
* don't add `distinct` because it can't generally have ordering and `@distinct`, we just know that it is **always** distinct without constraints;
* add `distinct` and start adding `distinct` to all `Capability` (trait from `type_`) to highlight when a capability is distinct. Creates redundancy, but it will be consistent.

I can also remove default `distinct` from other capabilities, but it will create additional risks!

Reply:
If we introduce a constraint by default, we are not forced to use it. If you don't see a need in it, then I'm happy to go without `distinct`ness for plays.
I was just thinking that `constraints` can be a tool for describing capabilities/types characteristics, and `plays` would just be always `distinct`. It's my issue to overgeneralize, that's why I am asking =)